### ISSUE TITLE
New rule request: Forbid comments in logical expressions or control statements

### ISSUE THREAD

**Issue by platinumazure** [eslint/eslint#5619]
Hi folks. I was inspired to write this request by going through the ESLint repo and fixing violations of `lines-around-comment`, which suggested a problematic pattern: Sometimes developers like to write a quick one- or two-liner conditional... and then a few more conditionals get added... and then one of them is complex enough to require a comment... Basically, my thought is, if part of your conditional is too complicated without a comment, the whole thing is too complicated without a comment and the whole thing should be extracted into a function.

It gets even better when you have this issue within an if or while statement, too...

## When the rules will warn. Include a description as well as sample code.

I'm not 100% sure on whether this should be restricted to LogicalExpressions only within test nodes of IfStatement or similar, or if it should just be done on any LogicalExpression.

If the former, here is an example where the rule would warn:

```js
if (condition1 &&
    doSomething().getValue() === "some value" &&
    // This is a really complicated condition
    EXPECTED_RESULTS.indexOf(doAnotherThing().result) !== -1
) {
    // code
}
```

If the latter, then it would also apply to LogicalExpressions that show up in AssignmentStatements, VariableDeclarators, or any other location.

## Whether the rule prevents an error or is stylistic.

It's stylistic, until you're actually trying to maintain those stupid conditionals. :laughing:  

## Why the rule should be in the core instead of creating a custom rule.

Part of the goal of linting is to call out potential problems, not just actual problems. Sometimes this means calling out an area of code to be refactored (in this case, into a function with good documentation). This concern is not library-specific, nor is the language syntax being evaluated more or less applicable to certain libraries.

## Are you willing to create the rule yourself?

Yes, although someone else might be able to get to it faster. :smile:

**Comment by platinumazure** [comment]
Another possibility might be checking number of conditionals or cyclomatic complexity per statement and flagging ones that are too complex. I should check to see if that's covered by `complexity`.

**Comment by alberto** [comment]
This is quite subjective. The same could be said of any non-doc comment.

**Comment by platinumazure** [comment]
Yes, I agree. Comments in a conditional are a symptom of it being too
complex, but there are better ways to measure it. Closing until I can come
up with a way to do that.
